### PR TITLE
Add documentation about asset design and behaviour

### PR DESCRIPTION
**Pull request by tunylund**
This pr is best reviewed by reading the document in this link

[documentation page](https://github.com/alphagov/whitehall/blob/asset-documentation/docs/assets.md)

[Trello](https://trello.com/c/KvwFO1F4/255-document-all-the-things)

### CODE REVIEW

File: docs/assets.md [hunk 1]
```diff
@@ -0,0 +1,149 @@
+# Asset Management in Whitehall
+
+*Assets, Attachments, Images, Logos, Featured Image, Lead Image, Response Form Data*
```
Comment (syed-ali-tw) [inline docs/assets.md:3]: can we say Response Form instead of Response Form Data?

File: docs/assets.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+Historically Whitehall had all file uploads stored within the Whitehall service. 
+Carrierwave was used to manage the http requests and related file storage within
+Whitehall.
+
+Later a separate service - Asset Manager - became the store for file uploads, taking ownership of every file uploaded.
+Carrierwave was still used to mange the http-request in Whitehall and a plentiful amount of configuration was
+written for carrierwave in order to make AssetManager behave as if it was just a file storage.
+These configurations rippled throughout Whitehall codebase and specialized its behaviour in
+Asset Manager, locking asset manager design and making the overhead of understanding Whitehall
+very heavy. Especially "Legacy url path" was a concept that heavily blocked any refactoring attempts.
```
Comment (syed-ali-tw) [inline docs/assets.md:25]: can we add a small description of what "legacy url path" is..
something on the lines: legacy url path was used like an asset-id. It was generated/formed inside whitehall and was based on model id of the asset + hardcoded path...

File: docs/assets.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+
+# Technical Design
+Most content types utilise attachments or images one way or other. 
+
+![Document types that use Attachments and Images](diagrams/asset_business_domains.png)
+
+![Model diagram](diagrams/asset_models.png)
+
+### Attachments upload flow
```
Comment (syed-ali-tw) [inline docs/assets.md:40]: do we also want to add a separate description for forms?

File: docs/assets.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+> File uploads are managed internally with Asynchronous background jobs. This helps the publisher-user
+have a fast user experience when uploading files. However it comes at the cost of complicated code flows
+that manage the successes and failures of the file uploads.
+
+# Business Requirements
+
+#### *Members of public* need to be see appropriate image sizes depending on the device they are using
+
+> Whitehall can publish several image urls as per necessary for various sizes.
```
Comment (syed-ali-tw) [inline docs/assets.md:129]: english is not my greatest strength, so I could be tottaly wrong here..
is as per necessary  = as per necessity is what we meant here?

File: docs/assets.md (hunk 3 above)
Comment (tunylund) [reply docs/assets.md:50]: i feel like it's unnecessary to include all the things - forms are effectively just file attachments, just like all the other things

File: docs/assets.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+# What is it
+
+There are several features in Whitehall that utilize Images or
+File Attachments.
+
+For example News Articles can have a list of attachments that
+provide further information about the news. Or a Person can have a picture
+which is shown on the person's page and also on any Speech page by that person.
+
+# How did it use to work
```
Comment (lauraghiorghisor-tw) [inline docs/assets.md:14]: `How it used to work` ?

File: docs/assets.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+Carrierwave was still used to mange the http-request in Whitehall and a plentiful amount of configuration was
+written for carrierwave in order to make AssetManager behave as if it was just a file storage.
+These configurations rippled throughout Whitehall codebase and specialized its behaviour in
+Asset Manager, locking asset manager design and making the overhead of understanding Whitehall
+very heavy. Especially "Legacy url path" was a concept that heavily blocked any refactoring attempts.
+
+"Legacy url path" was Whitehall defining what the url of an asset is, even though Asset Manager would actually
+own the asset. Other publishing applications would let Asset Manager define the url for assets.
+
+Later, the "Legacy url path" was removed and Whitehall Asset references became just simple rows in the
```
Comment (lauraghiorghisor-tw) [inline docs/assets.md:30]: This is us, right? "At the time of this writing" / "a tech debt team addressed this issue"
There is no immediate connection between the legacy bit and the ADR

**Pull request merged**